portfolio/src/main/java/com/google/sps/servlets/PageVisitsProcessor.java

```diff
@@ -0,0 +1,181 @@
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.gson.Gson;
+
+import javax.servlet.http.HttpServletRequest;
+import java.time.LocalDate;
+import java.time.ZoneOffset;
+import java.util.Calendar;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+public class PageVisitsProcessor {
+
+  private static final int MAX_DAYS = 30;
+  private static final String VISITS_TODAY = "Visits Today";
+
+  // Define constants for "Visits Today" entity properties
+  private static final String VISITS = "visits";
+  private static final String DATE = "date";
+  private static final String DAY_OF_WEEK = "dayOfWeek";
+  private static final String TIMESTAMP = "timestamp";
+
+  // Define constants for chart options (an enum might be a better choice)
+  private static final int CHART_BY_DATE = 0;
+  private static final int CHART_BY_WEEKDAY = 1;
+
+  private final DatastoreService datastore;
+
+  PageVisitsProcessor() {
+    this.datastore = DatastoreServiceFactory.getDatastoreService();
+  }
+
+  public String getDataJson(HttpServletRequest request) {
+    int choice = processGetRequest(request);
+    String chartDataJson;
+
+    if (choice == CHART_BY_WEEKDAY) {
+      chartDataJson = getDayOfWeekDataJson();
+    } else {
+      chartDataJson = getDateDataJson();
+    }
+
+    return chartDataJson;
+  }
+
+  private String getDayOfWeekDataJson() {
+    List<Entity> visitEntities = extractEntitiesByDate();
+    Map<Integer, Long> visitsByWeekday = mapVisitsPerWeekDay(visitEntities);
+
+    Gson gson = new Gson();
+    return gson.toJson(visitsByWeekday);
+  }
+
+  private String getDateDataJson() {
+    List<Entity> visitEntities = extractEntitiesByDate();
+    Map<LocalDate, Long> visitsByDate = mapVisitsPerDay(visitEntities);
+
+    Gson gson = new Gson();
+    return gson.toJson(visitsByDate);
+  }
+
+  private int processGetRequest(HttpServletRequest request) {
+    String chartChoice = request.getParameter("chart-choice");
+    int chartChoiceInt = CHART_BY_DATE;
+    try {
+      chartChoiceInt = Integer.parseInt(chartChoice);
+    } catch(NumberFormatException e) {
+      System.err.println("Could not convert to int: " + chartChoice);
+    }
+
+    return chartChoiceInt;
+  }
+
+  private Map<Integer, Long> mapVisitsPerWeekDay(List<Entity> visitEntities) {
+
```

Comment:
Remove blank line at the beginning of the method.

Reply:
Done.